### PR TITLE
Adjust altitude to compensate for FS2020/XP12 realistic altimetry model

### LINKED ISSUE #169: Adjust true altitude to compensate for MSFS/XP12 realistic altimetry model
> MSFS 2020 has an altimetry model that is more accurate/realistic than that of other sims, in that it simulates the effect of temperature on the pressure lapse rate in the atmosphere. Because of this, the indicated altitude that the pilot sees may be different from the true altitude of the aircraft. Since ATC clients show the reported true altitude in the data block when the aircraft is below the transition level, and since pilot clients render other aircraft at their reported true altitude, this creates an obvious problem. The more the temperature differs from a standard atmosphere, the greater the error.

And when X-Plane 12 is released, it will include a similar feature.

See the [proposed solution by Ross](https://github.com/swift-project/pilotclient/files/9012787/MSFS_Altitude_Fix_Proposal.pdf).

### PR DESCRIPTION
**Pull request by oktal3700**
Fixes #169

### CODE REVIEW

File: src/plugins/simulator/fsxcommon/simulatorfsxcommon.cpp [hunk 1]
```diff
@@ -1998,7 +2003,13 @@ namespace BlackSimPlugin::FsxCommon
                 // update situation
                 if (forceUpdate || !this->isEqualLastSent(result.getInterpolatedSituation()))
                 {
-                    SIMCONNECT_DATA_INITPOSITION position = this->aircraftSituationToFsxPosition(result, sendGround);
+                    // adjust altitude to compensate for FS2020 temperature effect
+                    CAircraftSituation situation = result;
+                    const CLength relativeAltitude = situation.geodeticHeight() - getOwnAircraftPosition().geodeticHeight();
+                    const double altitudeDeltaWeight = qBound(3000.0, relativeAltitude.abs().value(CLengthUnit::ft()), 6000.0) / 3000 - 1;
```
Comment (ltoenning) [inline src/plugins/simulator/fsxcommon/simulatorfsxcommon.cpp:2009]: This weight should be 1 for the ``relativeAltitude`` between 0-3000ft (fully apply offset) and linearly fade out till 6000ft.
If I am not mistaken it seems that the weight is 1 for 6.000ft only and decreases to reach 0 at 3000 ft.

File: src/plugins/simulator/xplane/simulatorxplane.cpp [hunk 2]
```diff
@@ -1076,7 +1088,13 @@ namespace BlackSimPlugin::XPlane
             const CInterpolationResult result = xplaneAircraft.getInterpolation(currentTimestamp, setup, aircraftNumber++);
             if (result.getInterpolationStatus().hasValidSituation())
             {
-                const CAircraftSituation interpolatedSituation(result);
+                CAircraftSituation interpolatedSituation(result);
+
+                // adjust altitude to compensate for XP12 temperature effect
+                const CLength relativeAltitude = interpolatedSituation.geodeticHeight() - getOwnAircraftPosition().geodeticHeight();
+                const double altitudeDeltaWeight = qBound(3000.0, relativeAltitude.abs().value(CLengthUnit::ft()), 6000.0) / 3000 - 1;
```
Comment (ltoenning) [inline src/plugins/simulator/xplane/simulatorxplane.cpp:1095]: Same as above:
This weight should be 1 for the relativeAltitude between 0-3000ft (fully apply offset) and linearly fade out till 6000ft.
If I am not mistaken it seems that the weight is 1 for 6.000ft only and decreases to reach 0 at 3000 ft.

File: src/plugins/simulator/fsxcommon/simconnectdatadefinition.cpp [hunk 3]
```diff
@@ -136,6 +136,8 @@ namespace BlackSimPlugin::FsxCommon
         hr +=  SimConnect_AddToDataDefinition(hSimConnect, CSimConnectDefinitions::DataOwnAircraft, "ROTATION VELOCITY BODY X", "Radians per second");
         hr +=  SimConnect_AddToDataDefinition(hSimConnect, CSimConnectDefinitions::DataOwnAircraft, "ROTATION VELOCITY BODY Y", "Radians per second");
         hr +=  SimConnect_AddToDataDefinition(hSimConnect, CSimConnectDefinitions::DataOwnAircraft, "ROTATION VELOCITY BODY Z", "Radians per second");
+        // FS2020
+        hr +=  SimConnect_AddToDataDefinition(hSimConnect, CSimConnectDefinitions::DataOwnAircraft, "INDICATED ALTITUDE CALIBRATED", "Feet");
```
[the remark this answers is not in the document]
Comment (oktal3700) [reply src/plugins/simulator/fsxcommon/simconnectdatadefinition.cpp:140]: That is what I presume

File: src/plugins/simulator/fsxcommon/simulatorfsxcommon.cpp (hunk 1 above)
Comment (oktal3700) [reply src/plugins/simulator/fsxcommon/simulatorfsxcommon.cpp:2009]: Good catch!

File: src/plugins/simulator/xplane/simulatorxplane.cpp (hunk 2 above)
Comment (oktal3700) [reply src/plugins/simulator/xplane/simulatorxplane.cpp:1095]: +1

**Review by ltoenning — APPROVED** [review]

**Pull request closed without merging**